source/extensions/common/aws/signers/iam_roles_anywhere_sigv4_signer_impl.cc

```diff
@@ -0,0 +1,85 @@
+#include "source/extensions/common/aws/signers/iam_roles_anywhere_sigv4_signer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Aws {
+
+std::string
+IAMRolesAnywhereSigV4Signer::createCredentialScope(absl::string_view short_date,
+                                                   absl::string_view override_region) const {
+  return fmt::format(IAMRolesAnywhereSigV4SignatureConstants::SigV4CredentialScopeFormat,
+                     short_date, override_region.empty() ? region_ : override_region,
+                     service_name_);
+}
+
+std::string IAMRolesAnywhereSigV4Signer::createStringToSign(
+    const X509Credentials x509_credentials, const absl::string_view canonical_request,
+    const absl::string_view long_date, const absl::string_view credential_scope) const {
+  auto& crypto_util = Envoy::Common::Crypto::UtilitySingleton::get();
+  if (x509_credentials.publicKeySignatureAlgorithm() ==
+      X509Credentials::PublicKeySignatureAlgorithm::RSA) {
+    return fmt::format(
+        IAMRolesAnywhereSigV4SignatureConstants::SigV4StringToSignFormat,
+        IAMRolesAnywhereSigV4SignatureConstants::X509SigV4RSA, long_date, credential_scope,
+        Hex::encode(crypto_util.getSha256Digest(Buffer::OwnedImpl(canonical_request))));
+  } else {
+    return fmt::format(
+        IAMRolesAnywhereSigV4SignatureConstants::SigV4StringToSignFormat,
+        IAMRolesAnywhereSigV4SignatureConstants::X509SigV4ECDSA, long_date, credential_scope,
+        Hex::encode(crypto_util.getSha256Digest(Buffer::OwnedImpl(canonical_request))));
+  }
+}
+
+std::string
```

Comment:
```suggestion
absl::StatusOr<std::string>
```

I would suggest returning `absl::Status` instead of `std::string` for operations that could fail and catch these failures in the caller method.

Reply:
The reason for this is because the majority of the signing code is duplicated from source/extensions/common/aws/signer_base_impl.cc, just slightly tweaked to use x509 credentials.
I had originally tried to refactor to avoid the duplication entirely but it just wasn't possible, and the duplication remains.
I will take another look over it to see if it makes sense to do this error handling.